### ISSUE TITLE
Bypass Backend caching when using AsyncResult

### ISSUE THREAD

**Issue by radulucaciu** [celery/celery#5107]
Hi,

this is more of a clarification issue, maybe I am missing something.

It seems that `AsyncResult._get_task_meta` method will always return results from `Backend`'s cache, as it calls `Backend.get_task_meta` without setting the cache param which defaults to true:
https://github.com/celery/celery/blob/master/celery/result.py#L410
https://github.com/celery/celery/blob/master/celery/backends/base.py#L360-L362

In my case, I am using Redis as the backend, and my results have a short TTL. 

The issues is that, even though the task has been expired from Redis, it is still returned from Celery's LRU cache, which is causing problems.

Is there any way to make sure that the result is not cached?

## Checklist

- [ ] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.

_Celery Version: celery==3.1.2_ 
## Steps to reproduce
- Use Redis as backend
- Create a new task
- Retrieve it as an AsyncResult
- Delete the key from Redis
- Try to list `AsyncResult.state` property

## Expected behavior
- AsyncResult.state to be `PENDING`

## Actual behavior
- AsyncResult.state is `SUCCESS` as it's returned from Celery's cache

**Comment by xirdneh** [comment]
Have you tried initializing the result by setting the cache as `False`?
```python
from celery.result import AsyncResult
res = Asyncresult(tas_id, cache=False)
res.get()
res.state
```

**Comment by radulucaciu** [comment]
@xirdneh I have not tried `AsyncResult(t_id, cache=False` but I doubt it will work since this is the `Backend`'s cache, not the `AsyncResult`'s. I will try it out in a few hours and report back if it worked, 

Wouldn't `forget()` remove it from Redis as well? In my particular scenario that is not the desired outcome, since I need to rely on Redis' TTL to persist the result for a while and only after that have them discarded. It's not clear to me from the [documentation](http://docs.celeryproject.org/en/latest/reference/celery.result.html#celery.result.AsyncResult.forget) if it removes it from Redis as well

**Comment by xirdneh** [comment]
@radulucaciu Sorry about the first comment, I was wrong about using `cache=False`.
The correct way to disable Celery's cache for a result is to `.forget()`.

When you do `AsyncResult(t_id).forget()` Celery will first [clear the result cache](https://github.com/celery/celery/blob/master/celery/result.py#L132) used in `AsyncResult`.
Then, `self.backend.forget(task_id)` is called which will [clear the backend cache.](https://github.com/celery/celery/blob/master/celery/backends/base.py#L326).
After that, the redis backend [will call 'self.result_consume.cancel_for(task_id)`](https://github.com/celery/celery/blob/master/celery/backends/redis.py#L303).
`cancel_for` [will disable any subscription](https://github.com/celery/celery/blob/master/celery/backends/redis.py#L135) to redis' queues.
Removing any subscription to Redis concerning that result will make Celery stop updating the result state, this is why the documentation say "possibly delete the result" because eventually the result will reach the Redis' TTL and be deleted by Celery. Celery should not explicitly delete the result when using `forget()`. I have tested this locally and the result is not deleted by Celery.

**Comment by radulucaciu** [comment]
Hi @xirdneh 

I tried calling `AsyncResult(t_id).forget()` and the result is deleted from redis.

```
my-redis-server:6379[1]> ttl my-task-meta-fd44b6266804554df3f4f3b859fd96b8
(integer) 7176
my-redis-server:6379[1]> time
1) "1539579695"
2) "940265"
```
After calling `forget()`:
```
my-redis-server:6379[1]> ttl my-task-meta-fd44b6266804554df3f4f3b859fd96b8
(integer) -2
my-redis-server:6379[1]> time
1) "1539579725"
2) "948928"
```

**Comment by bpereto** [comment]
Hm I think there is a implementation needed for that.
The functions exist (ex. https://github.com/celery/celery/blob/master/celery/backends/base.py#L371),
But no connection between AsyncResult and Backend.

**Comment by xirdneh** [comment]
I'm not sure why Redis is deleting the result I don't think celery actually sends a delete command to redis.
But yeah, probably @bpereto's solution would be the way to go.
```python
>>> app.backend.realod_task_result(task_id)
```

**Comment by bpereto** [comment]
Chain of calls:
* `AsyncResult.forget()` 
* `self.backend.forget(self.id)`
* `BaseKeyValueStoreBackend.forget(task_id)`
* `BaseKeyValueStoreBackend._forget(task_id)`
* `self.delete(self.get_key_for_task(task_id))`
This is for all backends whose inherit from  `BaseKeyValueStoreBackend`.

So yes, calling `result.forget()` does delete the result from the backend.

**Comment by xirdneh** [comment]
Ah, I didn't see that `self.delete`. Might be good to either expose the `reload_task_result` backend method to the result API or add a `delete` flag when forgetting the result.
I think exposing `reload_task_result` is better for readability.
